### ISSUE TITLE
"Replaying flow from last checkpoint" error

### ISSUE THREAD

**Issue by adelrustum** [corda/corda#5593]
Hi,

I posted this issue to stackoverflow earlier and Roger Willis had a look and suggested (on Slack) that it seems like a Corda issue.
Here's the link that has all details: https://stackoverflow.com/questions/58329416/replaying-flow-from-last-checkpoint-error

Thank you in advance.

**Comment by adelrustum** [comment]
Hello, any updates regarding this issue? I can provide more information if needed.

**Comment by adelrustum** [comment]
Here's the error:
```
net.corda.node.services.statemachine.FlowTimeoutException: replaying flow from the last checkpoint
    at net.corda.node.services.statemachine.SingleThreadedStateMachineManager$scheduleTimeoutException$$inlined$with$lambda$1.run(SingleThreadedStateMachineManager.kt:638) ~[corda-node-4.1.jar:?]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_222]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_222]
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_222]
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_222]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_222]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_222]
    at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
```

**Comment by adelrustum** [comment]
I attached IntelliJ to my node to debug remotely and I think the problem is happening in `NotaryFlow.call()` around the `val response = notarise(notaryParty)`.

**Comment by adelrustum** [comment]
Here's the call stack from the debugger:
```
call:62, NotaryFlow$Client (net.corda.core.flows)
call:36, NotaryFlow$Client (net.corda.core.flows)
subFlow:330, FlowStateMachineImpl (net.corda.node.services.statemachine)
subFlow:320, FlowLogic (net.corda.core.flows)
notariseAndRecord:198, FinalityFlow (net.corda.core.flows)
call:147, FinalityFlow (net.corda.core.flows)
call:39, FinalityFlow (net.corda.core.flows)
subFlow:330, FlowStateMachineImpl (net.corda.node.services.statemachine)
subFlow:320, FlowLogic (net.corda.core.flows)
call:75, ObserverAwareFinalityFlow (com.r3.corda.lib.tokens.workflows.internal.flows.finality)
call:35, ObserverAwareFinalityFlow (com.r3.corda.lib.tokens.workflows.internal.flows.finality)
subFlow:330, FlowStateMachineImpl (net.corda.node.services.statemachine)
subFlow:320, FlowLogic (net.corda.core.flows)
call:84, IssueTokensFlow (com.r3.corda.lib.tokens.workflows.flows.issue)
call:46, IssueTokensFlow (com.r3.corda.lib.tokens.workflows.flows.issue)
subFlow:330, FlowStateMachineImpl (net.corda.node.services.statemachine)
subFlow:320, FlowLogic (net.corda.core.flows)
call:37, IssueTokens (com.r3.corda.lib.tokens.workflows.flows.rpc)
call:26, IssueTokens (com.r3.corda.lib.tokens.workflows.flows.rpc)
subFlow:330, FlowStateMachineImpl (net.corda.node.services.statemachine)
subFlow:320, FlowLogic (net.corda.core.flows)
call:176, MyFlow (com.xxx.xxx.xxx.xxx.flows)
call:30, MyFlow (com.xxx.xxx.xxx.xxx.flows)
run:270, FlowStateMachineImpl (net.corda.node.services.statemachine)
run:46, FlowStateMachineImpl (net.corda.node.services.statemachine)
run1:1092, Fiber (co.paralleluniverse.fibers)
exec:788, Fiber (co.paralleluniverse.fibers)
doExec:100, RunnableFiberTask (co.paralleluniverse.fibers)
run:91, RunnableFiberTask (co.paralleluniverse.fibers)
call:511, Executors$RunnableAdapter (java.util.concurrent)
run:266, FutureTask (java.util.concurrent)
access$201:180, ScheduledThreadPoolExecutor$ScheduledFutureTask (java.util.concurrent)
run:293, ScheduledThreadPoolExecutor$ScheduledFutureTask (java.util.concurrent)
runWorker:1149, ThreadPoolExecutor (java.util.concurrent)
run:624, ThreadPoolExecutor$Worker (java.util.concurrent)
run:63, AffinityExecutor$ServiceAffinityExecutor$1$thread$1 (net.corda.node.utilities)
```

**Comment by adelrustum** [comment]
To remove any doubt that this problem might come from a custom code, I used the Java template (token-template branch, I updated tokens version to `1.1-RC01` and Corda OS to `4.3-RC01`) and deployed it to my node on GCP; I ran the flows that are included in this file: https://github.com/corda/cordapp-template-java/blob/token-template/workflows/src/main/java/com/template/flows/ExampleFlowWithEvolvableToken.java.  
So at first I ran `CreateEvolvableTokenFlow` then I used the Id of the resulting EvolvableToken to run `IssueEvolvableTokenFlow` which caused the same error: `FlowTimeoutException: replaying flow from the last checkpoint`.

**Comment by roger-that-dev** [comment]
Is there no other exception in the logs?

**Comment by adelrustum** [comment]
Here's the log from node start until the error happens:
[node-instance-2.log](https://github.com/corda/corda/files/3798492/node-instance-2.log)

**Comment by adelrustum** [comment]
Any updates guys? QA team is unable to do any work because of this problem.

**Comment by rick-r3** [comment]
That message occurs if the client node does not perceive a response from a clustered notary within a certain timeframe.  The message is expected if there are connectivity problems.  The node will then retry notarisation with another member of the notary cluster (and potentially repeat if that also has problems).  Does your cloud image have connectivity problems.  Can it notarise anything ?

**Comment by adelrustum** [comment]
So I do the following:
1. Create EvolvableTokenType (using the SDK's flow `CreateEvolvableTokens`), I don't get any errors here. This flow calls `ObserverAwareFinalityFlow` which in turn calls `FinalityFlow`, so I guess there is notarization happening here.
2. Access PostGres to get the linear Id of the above new state.
3. Run the issue flow that issues EvolvableTokens of the above new type. This flow throws the error.

**Comment by rick-r3** [comment]
I think step 1 does not involve the notary.  So step 3 is the first one to try and talk to the notary, and I think it is still plausible that the communication between node and notary is not working.

**Comment by adelrustum** [comment]
I will test the connection in the next 24 hours; sorry about the delay in response.

**Comment by adelrustum** [comment]
Rick, where do I get the IP address of the notary; or how do I test the connectivity to it? My node on GCP is part of the Corda TestNet and it uses its notary.

**Comment by adelRestom** [comment]
I get this when I try running the Node Explorer:

![image](https://user-images.githubusercontent.com/37046080/69764349-ace74d00-1135-11ea-9b19-7d0b9ca2a3d8.png)

I have all the required ports open:

![image](https://user-images.githubusercontent.com/37046080/69764427-e324cc80-1135-11ea-85c3-77eae9d7fa49.png)

Also the network:

![image](https://user-images.githubusercontent.com/37046080/69764486-0d768a00-1136-11ea-9f62-e9d15ad68a8a.png)

**Comment by adelrustum** [comment]
I was able to use node explorer with Finance cordapp on my node; I also issued USD to some node on the TestNet:

![image](https://user-images.githubusercontent.com/37046080/69908088-14a4ce80-13a8-11ea-9d8f-8dda7c18fd21.png)

**Comment by adelrustum** [comment]
Actually when I check the logs, I see things like this (I chose a random node when I issued some USD):
```
[INFO ] 2019-12-01T01:39:48,659Z [nioEventLoopGroup-2-3] netty.AMQPClient. - Retry connect to 172.39.2.168:10002
[INFO ] 2019-12-01T01:39:48,759Z [nioEventLoopGroup-2-4] netty.AMQPClient. - Retry connect to 172.39.2.99:10002
[INFO ] 2019-12-01T01:40:12,603Z [pool-11-thread-1] statemachine.FlowMonitor. - Flow with id 1cecb222-33eb-4a7d-97bd-57871d0ff136 has been waiting for 630 seconds to receive messages from parties [CN=, OU=18beffa3-a563-4916-b82e-34afda39d99f, O=TESTNET_TESTNET_Alphasystems-bc307, L=London, ST=, C=GB].
[INFO ] 2019-12-01T01:40:18,665Z [nioEventLoopGroup-2-5] netty.AMQPClient. - Failed to connect to 172.39.2.168:10002
[INFO ] 2019-12-01T01:40:18,793Z [nioEventLoopGroup-2-6] netty.AMQPClient. - Failed to connect to 172.39.2.99:10002
[INFO ] 2019-12-01T01:41:12,603Z [pool-11-thread-1] statemachine.FlowMonitor. - Flow with id 1cecb222-33eb-4a7d-97bd-57871d0ff136 has been waiting for 690 seconds to receive messages from parties [CN=, OU=18beffa3-a563-4916-b82e-34afda39d99f, O=TESTNET_TESTNET_Alphasystems-bc307, L=London, ST=, C=GB].
```

**Comment by adelrustum** [comment]
This is my `node.conf`, for some reason the `node.conf` that's generated by the script that comes form TestNet uses the old format (keys surrounded by double quotes which is deprecated); `compatibilityZoneURL` is deprecated as well so I don't know why the script uses that. Anyway, I changed the file to the new format (I'm also using Corda 4.3):
```
compatibilityZoneURL = "https://netmap.testnet.r3.com"
emailAddress = "xxx"
jarDirs = [ "plugins", "cordapps" ]
myLegalName = "OU=xxx, O=TESTNET_xxx, L=London, C=GB"
keyStorePassword = "xxx"
trustStorePassword = "xxx"
crlCheckSoftFail = true
database = {
    transactionIsolationLevel = "READ_COMMITTED"
    initialiseSchema = "true"
}
dataSourceProperties {
    dataSourceClassName = "org.postgresql.ds.PGSimpleDataSource"
    dataSource.url = "jdbc:postgresql://xxx/postgres"
    dataSource.user = xxx
    dataSource.password = xxx
}
p2pAddress = "xxx:10002"
rpcSettings {
    useSsl = false
    standAloneBroker = false
    address = "0.0.0.0:10003"
    adminAddress = "0.0.0.0:10004"
}
rpcUsers = [
    { 
	username=xxx, 
	password=xx, 
	permissions=[ 
		"ALL" 
	] }
]
devMode = false
cordappSignerKeyFingerprintBlacklist = []
useTestClock = false
```

**Comment by roastario** [comment]
what is the firewall setup on your machine?

**Comment by adelrustum** [comment]
@roastario 

![image](https://user-images.githubusercontent.com/37046080/69963442-d86f8c00-14d5-11ea-8617-56ac01d558ff.png)

![image](https://user-images.githubusercontent.com/37046080/69963680-5a5fb500-14d6-11ea-9aa8-499469ef8c96.png)

**Comment by adelrustum** [comment]
Do I need to change anything here?

![image](https://user-images.githubusercontent.com/37046080/69969401-f80cb180-14e1-11ea-9cc3-b5e115020f8f.png)

**Comment by adelrustum** [comment]
I debugged this with Fuzz and during the run of the script that comes from TestNet it proposes some IP address (I don't know where it gets it from), I used to say "yes", Fuzz suggested saying "no" and supplying the VM external IP address which I did: 
![image](https://user-images.githubusercontent.com/37046080/69982425-167ea700-14fa-11ea-833c-0d1e14cf88fa.png)

So, I rebuilt my node and retried all the steps. Same error :-/
Here's the full log file of the node:
[node-instance-1.log](https://github.com/corda/corda/files/3912577/node-instance-1.log)

Btw, I'm able to telnet to my VM external IP with ports 10002 and 10003.

**Comment by adelrustum** [comment]
From Corda's code documentation:
```
* ``FlowTimeoutException``:
     This is used internally by the notary client flow when talking to an HA notary.  It’s used to cause the client to try and talk to a different
     member of the notary cluster if it doesn't hear back from the original member it sent the request to within a “reasonable” time.
     The time is hard to document as the notary members, if actually alive, will inform the requester of the ETA of a response.
     This can occur an infinite number of times.  i.e. we never give up notarising.  No intervention required.
```

**Comment by adelrustum** [comment]
Can someone from R3 create a node on GCP and try one of their CorDapps that requires notarization?

**Comment by adelrustum** [comment]
The script was suggesting the wrong IP during the below step, I followed Fuzz suggestion and said "No" and used my VM external IP (instead of what's suggested); initially it didn't work, but after an hour and a half it did. Not sure what happened in between because I didn't change anything in my configuration.

![image](https://user-images.githubusercontent.com/37046080/70000423-a1729800-1520-11ea-8627-c1ba684b7864.png)

**Comment by UmeshPandit** [comment]
what is correct fix w.r.t to Azure if any.

**Comment by adelrustum** [comment]
Umesh, like I mentioned in my last comment, upon running the script from TestNet at one point it would suggest an IP address (see last printscreen), I used to blindfully say "yes" and accept an incorrect IP, instead you must say "no" and provide the external IP of your VM, if you provide the wrong IP; the notary won't recognize calls coming form your correct IP and reject them, so your flow is hanging waiting on a reply from the notary and eventually it times out and goes to flow hospital.
In my case to fix the issue, I re-ran the installation script (i.e. I redeployed my node).
